cli_tools_e2e_test/gce_windows_upgrade/test_suites/windows_upgrade/windows_upgrade_test_suite.go

```diff
@@ -0,0 +1,448 @@
+	}, logger, testCase) {
+		return
+	}
+
+	// attach data disk
+	for dataDiskIndex := 1; dataDiskIndex <= dataDiskCount; dataDiskIndex++ {
+		diskName := fmt.Sprintf("%v-%v", instanceName, dataDiskIndex)
+		if !utils.RunTestCommand("gcloud", []string{
+			"compute", "disks", "create", fmt.Sprintf("--zone=%v", testProjectConfig.TestZone),
+			fmt.Sprintf("--project=%v", testProjectConfig.TestProjectID), "--size=10gb",
+			"--image=projects/compute-image-tools-test/global/images/empty-ntfs-10g",
+			diskName,
+		}, logger, testCase) {
+			return
+		}
+
+		if !utils.RunTestCommand("gcloud", []string{
+			"compute", "instances", "attach-disk", instanceName, fmt.Sprintf("--zone=%v", testProjectConfig.TestZone),
+			fmt.Sprintf("--project=%v", testProjectConfig.TestProjectID),
+			fmt.Sprintf("--disk=%v", diskName),
+		}, logger, testCase) {
+			return
+		}
+	}
+
+	// set original startup script
+	if expectedScriptURL != "" {
+		key := "windows-startup-script-url"
+		if expectedScriptURL == "original-backup" {
+			key = "windows-startup-script-url-backup"
+		}
+		_, err := computeUtils.SetMetadata(ctx, testProjectConfig.TestProjectID, testProjectConfig.TestZone, instanceName,
+			key, expectedScriptURL, true)
+		if err != nil {
+			utils.Failure(testCase, logger, fmt.Sprintf("Failed to set metadata for %v: %v", instanceName, err))
+			return
+		}
+	}
+
+	var success bool
+	if testType == utils.Wrapper {
+		cmd := utils.RunTestCommandAsync(cmd, args, logger, testCase)
+
+		go func() {
+			// send a INT signal to fail the upgrade
+			if triggerFailure {
+				// wait for "preparation" to finish
+				instance, err := computeUtils.CreateInstanceObject(ctx, testProjectConfig.TestProjectID, testProjectConfig.TestZone, instanceName, true)
+				if err != nil {
+					utils.Failure(testCase, logger, fmt.Sprintf("Failed to fetch instance object for %v: %v", instanceName, err))
+					return
+				}
+				expectedOutput := "Beginning upgrade startup script."
+				logger.Printf("[%v] Waiting for `%v` in instance serial console.", instanceName,
+					expectedOutput)
+				if err := instance.WaitForSerialOutput(
+					expectedOutput, 1, 15*time.Second, 20*time.Minute); err != nil {
+					testCase.WriteFailure("Error during waiting for preparation finished: %v", err)
+					return
+				}
+
+				err = cmd.Process.Signal(os.Interrupt)
+				if err != nil {
+					utils.Failure(testCase, logger, fmt.Sprintf("Failed to send ctrl+c to upgrade %v: %v", instanceName, err))
+					return
+				}
+			}
+		}()
+
+		err := cmd.Wait()
+		if err != nil {
+			success = false
+		} else {
+			success = true
+		}
+	} else {
+		success = utils.RunTestForTestType(cmd, args, testType, logger, testCase)
+	}
+
+	verifyUpgradedInstance(ctx, logger, testCase, testProjectConfig, instanceName, success,
+		expectSuccess, expectedScriptURL, autoRollback, dataDiskCount, expectValidationFailure)
+}
+
+func verifyUpgradedInstance(ctx context.Context, logger *log.Logger, testCase *junitxml.TestCase,
```

Comment:
This func is very long (110+ lines). Consider extracting portions of it to separate functions, like verifyUpgradedInstanceDisks, verifyUpgradedInstanceLicenses,verifyUpgradedInstanceOnSuccess, verifyUpgradedInstanceOnFailure or similar.

Reply:
done